Docs: rewrite mobile OTA and Dev Tool documentation (#9061)
Couchers-org/couchers: app/mobile/README.md

# Couchers Mobile App

React Native mobile app built with [Expo](https://expo.dev). We maintain three separate app variants (staging, production, and a development tool) that can coexist on the same device with different bundle IDs and API endpoints.

> You don't need to build the app locally: the prebuilt **Couchers Dev Tool** variant covers almost all mobile dev work.

Reviewer: What's the use case for when you cannot use it for mobile dev?

Author: it won't work if you are upgrading certain dependencies (like expo itself). this is discussed later in the readme.

_Readme last updated: 2026/06/13._

## Table of Contents

1. [Quick Start](#1-quick-start)
2. [Full local development setup](#2-full-local-development-setup)
3. [Local native builds](#3-local-native-builds)
4. [Releases](#4-releases)
5. [App variants](#5-app-variants)
6. [Updating dependencies](#6-updating-dependencies)
7. [Learn more](#7-learn-more)

## 1. Quick Start

You can do local development without having to build the mobile app from scratch using the Couchers Dev Tool. Download it for Android or the iOS simulator from the [dev tool builds page](https://develop--devtool-builds.preview.couchershq.org/). To get it on your physical iPhone, ask a mobile dev lead to invite you via TestFlight.

On your dev machine, you need `nodejs` v22. We recommend using `nvm` (the [node version manager](https://github.com/nvm-sh/nvm)) to do this. You can install it with:

```sh

> curl -sL https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash

Reviewer: Can you make this cleaner and reference a named script in package.json? That's the standard. So move the longer command there under "scripts" and have the command name what it does.

Also add a line for what this does. I see a version number here will this get out of date quickly?

Author: This can't be included as a script to be run by npm since this runs before npm is installed. It explains on the line above what it's doing. This is just the package manager, it should not need to be updated regularly.

Btw this line is copied from the web readme.

```

> **You will need to restart your terminal before `nvm` becomes available.** (See the [web frontend readme](../web/readme.md) for a deeper discussion on `nvm` and deps).

Reviewer: This line seems kind of random, does it belong here? Should that be under some kind of "Initial Setup" or "before you begin" section?

Author: what do you mean? yes this is just the initial setup. it's similar to web readme

Now run the following commands to get up and running:

```sh
## Check out the repo and navigate to app/mobile

> git clone https://github.com/Couchers-org/couchers.git

Reviewer: I think there should be a comment "if you haven't ever cloned the github repo before for web, backend, etc."

Author: again, this is the same philosophy as the web readme: it's a few-line quick start that someone can just plug along with. if you already have things set up and are familiar with things (e.g. existing dev), it's pretty clear what everything is doing.

cd couchers/app/mobile

# install and use the right node version
nvm install
# install dependencies
npm install
# generate the proto stubs
npm run build:protos

# start the dev server through the dev tool
npm run start:devtool
```

> **Open the Dev Tool app and scan the QR code** (your phone and computer must be on the same network). TypeScript changes hot-reload automatically.

Reviewer: I think you scan it with your camera not the dev tool app. At least I don't see a QR code scanning section in thee devtool?

Author: thanks for catching that, will update.

You should now see the mobile app on your phone.

> This setup will point at your local version of the React Native app, but it will point at the staging web app and backend. You probably want to follow the instructions on local dev below to run the full stack locally.

Reviewer: ```suggestion
This setup will point at your local version of the React Native app with the staging web app and backend. If you change any code within the web or backend folders, you will need to run the full stack locally to see your changes. Follow the instructions on local dev below.
```

## 2. Full local development setup

To actually do full local development, you will want to run the entire stack on your dev machine, and point the dev tool at that instead of the staging environment.

> First, run the following script which patches environment variables to point the app at your local running versions of the frontend and backend. It auto-detects your IP and updates all config files at once:

Reviewer: ```suggestion
To run the mobile, frontend and backend environments fully locally, you need to run the entire stack on your dev machine, and point the dev tool at that instead of the staging environment.

First, run the following script which patches environment variables to point the app at your local running versions of the frontend and backend. It auto-detects your IP and updates all config files at once:
```

```bash
npm run setup:local
```

On a first run: run each of the quick starts in the `backend/` and `web/` folders to get your machine set up.

Now run everything from the repo root, each in a separate terminal:

```bash

> # Terminal 1: backend (requires Docker)

Reviewer: ```suggestion
# Terminal 1: backend (requires Docker)

``` bash
cd app/backend

make protos
```

docker compose up --build

# Terminal 2: web frontend
cd app/web && yarn start

> 

Reviewer: I think we should do:

cd app/web
yarn install
yarn start

> # Terminal 3: Expo

Reviewer: ```suggestion
# Terminal 3: Expo (Mobile)
```

### Android Emulator

1. Launch an emulator from Android Studio (AVD Manager)
2. Run:
```bash
npx expo run:android
npm run setup:local:restore
```

See [Expo Android emulator docs](https://docs.expo.dev/workflow/android-studio-emulator/) for troubleshooting.

## 4. Releasing Staging Build
### Checks

Staging builds point to `dev-api.couchershq.org` and are for internal testing only. The staging app (`Couchers (Staging)`) can be installed alongside the production app.
Run these before pushing (CI runs them too):

> **CI builds staging automatically.** On every `develop` push, CI cuts and submits a fresh staging store build when the native fingerprint changes (and ships pure JS/TS changes over the air), exactly like production — see `app/.gitlab-ci.yml` (`build:`/`deploy:mobile-native-staging-*` and `build:mobile-ota-staging`). The manual commands below are the ad-hoc fallback; set `FORCE_NATIVE_BUILD_AND_SUBMIT=true` on a pipeline to force a fresh store build when the fingerprint is unchanged.

### Before Building

Run these checks:
```bash
```sh
npm run format    # auto-fix lint errors + format code
npm run lint      # check for remaining lint errors
npm test          # run tests

> npx expo-doctor   # verify build configuration

Reviewer: We still want `npx expo-doctor` somewhere as this is more about making sure we have the right package versions. It will fail on the build if not but then that wastes time.

Can add it to a "before doing a build or opening a PR" step if you prefer.

Author: i put this back, thanks!

> Only needed when developing native changes (when Expo fingerprints would change).

Reviewer: Maybe add some common examples here? When installing or removing a new package, etc?

Author: added

> If you hit iOS signing issues, open `app/mobile/ios` in Xcode and set up signing there.

Reviewer: Nice to have, link the expo docs for this as I've had to do it more than once and it's always a bit of a quest at first.

Author: ohhh i didn't realize they had an inof page, added a link

> All release builds (Dev Tool, Stage, Prod) are done in CI ([`app/.gitlab-ci.yml`](../.gitlab-ci.yml)) on every push to `develop`, for both staging and production:

Reviewer: What if we want to release to staging before we do a prod build?

Author: can use the force flags; also we can do a prod build without releasing it, so should be harmless

Reviewer: is this documented how to release only to staging somewhere?

Author: i added the instructions on how to set the env vars for gitlab build for this

> - **Native fingerprint changed** → CI builds a fresh store build on EAS and submits it: iOS to TestFlight, Android to the Play internal testing track. (Set `FORCE_NATIVE_BUILD_AND_SUBMIT=true` on a pipeline to force a build when the fingerprint is unchanged.)

Reviewer: How do we "Set `FORCE_NATIVE_BUILD_AND_SUBMIT=true` on a pipeline"?

Author: thanks, added instructions

> - **Fingerprint unchanged** (JS/TS-only changes) → CI ships the changes as a signed [over-the-air update](../../docs/native-ota-updates.md) instead — no store build needed.

Reviewer: So we just have to merge to develop and that's it?

Author: yep

> **Staging** builds are for internal testing and are available as soon as CI submits them (TestFlight after Apple's automated processing; Play internal testing immediately). Optionally add release notes in [App Store Connect](https://appstoreconnect.apple.com) (Couchers (Staging) → TestFlight → build → "What to Test") or the [Google Play Console](https://play.google.com/console) (Couchers (Staging) → Testing → Internal testing → release notes).

Reviewer: ```suggestion
**Staging** builds are for internal testing and are available as soon as CI submits them (TestFlight after Apple's automated processing; Play internal testing immediately). Ideally add release notes in [App Store Connect](https://appstoreconnect.apple.com) (Couchers (Staging) → TestFlight → build → "What to Test") or the [Google Play Console](https://play.google.com/console) (Couchers (Staging) → Testing → Internal testing → release notes) so testers know what to look for.

Update `RELEASE_NOTES.md` as you code to keep track of changes then copy/paste this after building.
```

| **Dev Tool**   | Couchers Dev Tool  | `org.couchers.devtool.ios` | `org.couchers.devtool.android` | `dev-api.couchershq.org`  |
| **Staging**    | Couchers (Staging) | `org.couchers.staging.ios` | `org.couchers.staging.android` | `dev-api.couchershq.org`  |
| **Production** | Couchers           | `org.couchers.ios`         | `org.couchers.android`         | `api.couchers.org`        |

Build profiles in `eas.json` set an `APP_VARIANT` environment variable, which `app.config.js` reads to configure bundle IDs, app names, and API endpoints. The Dev Tool is a [development build](https://docs.expo.dev/develop/development-builds/introduction/) — a developer utility, not a release flavor — see [docs/native-dev-tool.md](../../docs/native-dev-tool.md).

## 6. Updating dependencies

Mobile dependencies are **not** managed by Dependabot. Expo requires specific compatible versions of packages tied to each SDK version, so update through Expo's tooling:

```sh
# Check for outdated or incompatible packages
npx expo install --check

# Auto-fix to Expo-compatible versions
npx expo install --fix

> ```

Reviewer: I think we can just do `npx expo-doctor`

Author: fixed
